src/chapters/aspdotnet-model-validation/studio.rst

```diff
@@ -0,0 +1,120 @@
+#.  Username is between 5 and 15 characters.
+#.  Email is optional.
+#.  If provided, the email has the format of a valid email address.
+#.  The password is between 6 and 20 characters.
+
+Refactoring ``UserController``
+------------------------------
+
+With our new ViewModel set up and ready, we need to refactor our action methods to make use of ``AddUserViewModel``.
+First, pass an instance of ``AddUserViewModel`` to ``User/Add.cshtml`` via our ``Add()`` action method.
+Next, to refactor the ``SubmitAddUserForm()`` action method, we need to do the following:
+
+#. Take in the instance of ``AddUserViewModel`` as a parameter.
+#. Use ``ModelState.IsValid`` to make sure that the conditions outlined using validation attributes have been met.
+#. If ``ModelState.IsValid`` is ``true``, we want to double check that the value of ``Password`` is equal to ``VerifyPassword``.
+   If it is, we can create an instance of ``User`` and pass it to the ``User/Index.cshtml`` view.
+   If the two passwords do not match, we need to reload the form with an error message displayed.
+
+Validating That Passwords Match
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Last studio, we added some validation checks to make sure the password fields match.
+Now we have two validation sections: one for the attribute-configured
+validation (which checks ``ModelState.IsValid``), and one that checks
+that the password fields match. Make sure they work in-sync with each
+other to properly return to the form if any of the validation conditions
+fail.
+
+.. admonition:: Tip
+
+   You can, in fact, validate that passwords match using attributes by
+   taking a slightly different approach than we’ve done here. We
+   outline how to do so in the Bonus Mission section.
+
+Refactoring the Views
+---------------------
+
+The final step in adding ViewModels and validation to our ``SpaDay`` application is to refactor our views!
+
+In ``User/Index.cshtml``:
+
+#. Make the ``User`` model accessible to the view.
+#. Replace instances of ``ViewBag`` with the ``@Model`` syntax.
+
+If we do this correctly, when a form with valid data is submitted, we should still see our personalized greeting!
+
+In ``User/Add.cshtml``:
+
+#. Make ``AddUserViewModel`` accessible to the view.
+#. Keep the ``@if`` statement up at the top so if the two passwords don't match, we can see the error message!
+#. Replace the inner contents of the form with a properly formatted form field for each property in the ViewModel.
+   Each of your form fields should look something like this:
+
+   .. sourcecode:: guess
+      :linenos:
+
+      <div class="form-group">
+         <label asp-for="PropertyName">PropertyName</label>
+         <input asp-for="PropertyName" />
+         <span asp-validation-for="PropertyName"></span>
+      </div>
+
+   Note when it comes to the password fields you may want to add a ``type`` attribute to the ``<input>`` tag so the users cannot see what they are typing!
```

Comment:
this is a nice note. instead of emphasizing that the user wont see what they're typing, we say emphasize that the password input type provides better data security

Reply:
Done!